File: algorithms/graph/prims/java/Prim.java
import java.util.*;

/**
 * Prim's algorithm to find the Minimum Spanning Tree (MST) total weight.
 * Uses a weighted adjacency list.
 */
public class Prim {
    public static int prim(int numVertices, Map<Integer, List<List<Integer>>> adjList) {
        boolean[] inMST = new boolean[numVertices];
        int[] key = new int[numVertices];
        Arrays.fill(key, Integer.MAX_VALUE);
        key[0] = 0;

        // Priority queue: [weight, vertex]
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, 0});

        int totalWeight = 0;

        while (!pq.isEmpty()) {
            int[] current = pq.poll();
            int w = current[0];
            int u = current[1];

            if (inMST[u]) continue;

            inMST[u] = true;
            totalWeight += w;

            List<List<Integer>> neighbors = adjList.getOrDefault(u, Collections.emptyList());
            for (List<Integer> edge : neighbors) {
                int v = edge.get(0);
                int weight = edge.get(1);
                if (!inMST[v] && weight < key[v]) {
                    key[v] = weight;
                    pq.offer(new int[]{weight, v});
                }
            }
        }

        return totalWeight;
    }

    public static void main(String[] args) {
        Map<Integer, List<List<Integer>>> adjList = new HashMap<>();
        adjList.put(0, Arrays.asList(Arrays.asList(1, 10), Arrays.asList(2, 6), Arrays.asList(3, 5)));
        adjList.put(1, Arrays.asList(Arrays.asList(0, 10), Arrays.asList(3, 15)));
        adjList.put(2, Arrays.asList(Arrays.asList(0, 6), Arrays.asList(3, 4)));
        adjList.put(3, Arrays.asList(Arrays.asList(0, 5), Arrays.asList(1, 15), Arrays.asList(2, 4)));

        int result = prim(4, adjList);
        System.out.println("MST total weight: " + result);
    }
}

File: algorithms/graph/prims/scala/Prim.scala
import scala.collection.mutable

/**
 * Prim's algorithm to find the Minimum Spanning Tree (MST) total weight.
 * Uses a weighted adjacency list.
 */
object Prim {
  def prim(numVertices: Int, adjList: Map[Int, List[(Int, Int)]]): Int = {
    val inMST = Array.fill(numVertices)(false)
    val key = Array.fill(numVertices)(Int.MaxValue)
    key(0) = 0

    // Priority queue: (weight, vertex)
    val pq = mutable.PriorityQueue[(Int, Int)]()(Ordering.by[(Int, Int), Int](-_._1))
    pq.enqueue((0, 0))

    var totalWeight = 0

    while (pq.nonEmpty) {
      val (w, u) = pq.dequeue()

      if (!inMST(u)) {
        inMST(u) = true
        totalWeight += w

        for ((v, weight) <- adjList.getOrElse(u, List.empty)) {
          if (!inMST(v) && weight < key(v)) {
            key(v) = weight
            pq.enqueue((weight, v))
          }
        }
      }
    }

    totalWeight
  }

  def main(args: Array[String]): Unit = {
    val adjList = Map(
      0 -> List((1, 10), (2, 6), (3, 5)),
      1 -> List((0, 10), (3, 15)),
      2 -> List((0, 6), (3, 4)),
      3 -> List((0, 5), (1, 15), (2, 4))
    )

    val result = prim(4, adjList)
    println(s"MST total weight: $result")
  }
}
